products/kms/terraform.yaml

```diff
@@ -45,7 +45,54 @@ overrides: !ruby/object:Overrides::ResourceOverrides
       custom_delete: templates/terraform/custom_delete/kms_key_ring.erb
       decoder: templates/terraform/decoders/kms_key_ring.go.erb
       encoder: templates/terraform/encoders/send_nil_body.go.erb
-      extra_schema_entry: templates/terraform/extra_schema_entry/kms_key_ring_self_link.erb
+      extra_schema_entry: templates/terraform/extra_schema_entry/kms_self_link.erb
+  CryptoKey: !ruby/object:Overrides::Terraform::ResourceOverride
+    description: |
+      {{description}}
+
+      ~> **Note:** CryptoKeys cannot be deleted from Google Cloud Platform.
+      Destroying a Terraform-managed CryptoKey will remove it from state
+      and delete all CryptoKeyVersions, rendering the key unusable, but *will
+      not delete the resource on the server.* When Terraform destroys these keys,
+      any data previously encrypted with these keys will be irrecoverable.
+      For this reason, it is strongly recommended that you add lifecycle hooks
+      to the resource to prevent accidental destruction.
+    id_format: "{{key_ring}}/cryptoKeys/{{name}}"
+    import_format: ["{{key_ring}}/cryptoKeys/{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "kms_crypto_key_basic"
+        primary_resource_id: "example-key"
+        skip_test: true
+      - !ruby/object:Provider::Terraform::Examples
+        name: "kms_crypto_key_assymmetric_sign"
+        primary_resource_id: "example-assymmetric-sign-key"
+        skip_test: true
+    docs: !ruby/object:Provider::Terraform::Docs
+      attributes: |
+        * `self_link`: The self link of the created CryptoKey. Its format is `{{key_ring}}/cryptoKeys/{{name}}`.
+    properties:
+      creationTime: !ruby/object:Overrides::Terraform::PropertyOverride
+        exclude: true
+      keyRing: !ruby/object:Overrides::Terraform::PropertyOverride
+        diff_suppress_func: 'kmsCryptoKeyRingsEquivalent'
+        ignore_read: true
+      name: !ruby/object:Overrides::Terraform::PropertyOverride
+        ignore_read: true
+      rotationPeriod: !ruby/object:Overrides::Terraform::PropertyOverride
+        validation: !ruby/object:Provider::Terraform::Validation
+          function: 'orEmpty(validateKmsCryptoKeyRotationPeriod)'
+      versionTemplate: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
+      versionTemplate.protectionLevel: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
```

Comment:
Why the change to this instead of a default?

Reply:
This does not need to be specified at creation time and the API will fill it with a default of `SOFTWARE`. Isn't that what `default_from_api` is for?

Reply:
Also, the defaults only apply to symmetric keys. Leaving it up to the API to fill the default value (for symmetric keys) make the error message more clear when you try to create an asymmetric key and do not specify a versionTemplate:
```
resource "google_kms_crypto_key" "my-key-4" {
  name            = "my-key-4"
  key_ring        = google_kms_key_ring.my_key_ring.self_link
  purpose  = "ASYMMETRIC_SIGN"

  lifecycle {
    prevent_destroy = true
  }
}
```
```
Terraform will perform the following actions:

  # google_kms_crypto_key.my-key-4 will be created
  + resource "google_kms_crypto_key" "my-key-4" {
      + id        = (known after apply)
      + key_ring  = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring"
      + name      = "my-key-4"
      + purpose   = "ASYMMETRIC_SIGN"
      + self_link = (known after apply)

      + version_template {
          + algorithm        = (known after apply)
          + protection_level = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```
then
```
google_kms_crypto_key.my-key-4: Creating...

Error: Error creating CryptoKey: googleapi: Error 400: Request field version_template.algorithm is required if purpose is not ENCRYPT_DECRYPT.

  on kms.tf line 49, in resource "google_kms_crypto_key" "my-key-4":
  49: resource "google_kms_crypto_key" "my-key-4" {
```
Otherwise you'd get an error saying that `GOOGLE_SYMMETRIC_ENCRYPTION`(which the user didn't explicitly specify) is not a valid algorithm for asymmetric keys. I find the error above more user friendly.

Reply:
Yes, but that's also what defaults are for.

There isn't any real guidance on when to use one vs the other. Defaults have the advantage of showing up in the plan (vs <computed>), and fields with defaults can also be changed by removing the field from configs (vs computed fields, which don't show a diff if it isn't defined). Computed fields have the advantage of reacting appropriately if the API changes what the default value is.

My opinion generally for this sort of thing is that we shouldn't change things that aren't broken, but I'm also not going to block approval on it.

Reply:
Thanks for clarifying. Makes sense to change the protectionLevel to "SOFTWARE" since that always, apply, but I'll leave the algorithm as default_from_api since the old hardcoded default no longer applies when we're dealing with asymmetric keys. Pushed the changed.

Reply:
Oh by the way, my comment got in a bit of a race condition with yours, so it should be read as having responded to your previous comment.

I'm a bit confused about your remarks regarding `algorithm`. Isn't that field marked as required?

Reply:
`algorithm` is required only if `versionTemplate` is specified, but `versionTemplate` is not required. See below an example:
```
POST /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys?alt=json&cryptoKeyId=my-key-5&prettyPrint=false HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: google-api-go-client/0.5 Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 30
Content-Type: application/json
Accept-Encoding: gzip

{
 "purpose": "ENCRYPT_DECRYPT"
}

-----------------------------------------------------
2019/06/13 00:36:23 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 22:36:23 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
 "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-key-5",
 "primary": {
  "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-key-5/cryptoKeyVersions/1",
  "state": "ENABLED",
  "createTime": "2019-06-12T22:36:23.559829707Z",
  "protectionLevel": "SOFTWARE",
  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION",
  "generateTime": "2019-06-12T22:36:23.559829707Z"
 },
 "purpose": "ENCRYPT_DECRYPT",
 "createTime": "2019-06-12T22:36:23.559829707Z",
 "versionTemplate": {
  "protectionLevel": "SOFTWARE",
  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION"
 }
}
-----------------------------------------------------
```